core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1201,142 +1081,170 @@ object ZManaged {
   }
 
   /**
-   * A `FinalizerRef` describes the finalizers associated with a scope.
-   *
+   * A finalizer used in a [[ReleaseMap]]. The [[Exit]] value passed to
+   * it is the result of executing [[ZResource#use]] or an arbitrary value
+   * passed into [[ReleaseMap#release]].
    */
-  trait FinalizerRef[R] {
+  type Finalizer = Exit[Any, Any] => UIO[Any]
+  object Finalizer {
+    val noop: Finalizer = _ => UIO.unit
+  }
+
+  final class IfM[R, E](private val b: ZManaged[R, E, Boolean]) extends AnyVal {
+    def apply[R1 <: R, E1 >: E, A](
+      onTrue: => ZManaged[R1, E1, A],
+      onFalse: => ZManaged[R1, E1, A]
+    ): ZManaged[R1, E1, A] =
+      b.flatMap(b => if (b) onTrue else onFalse)
+  }
+
+  trait PreallocationScope {
+    def apply[R, E, A](managed: ZManaged[R, E, A]): ZIO[R, E, Managed[Nothing, A]]
+  }
+
+  final class ProvideSomeLayer[R0 <: Has[_], -R, +E, +A](private val self: ZManaged[R, E, A]) extends AnyVal {
+    def apply[E1 >: E, R1 <: Has[_]](
+      layer: ZLayer[R0, E1, R1]
+    )(implicit ev1: R0 with R1 <:< R, ev2: NeedsEnv[R], tagged: Tagged[R1]): ZManaged[R0, E1, A] =
+      self.provideLayer[E1, R0, R0 with R1](ZLayer.identity[R0] ++ layer)
+  }
+
+  /**
+   * A key used for identifying finalizers in a [[ReleaseMap]].
+   */
+  case class ReleaseKey(l: Long) extends AnyVal {
```

Comment:
Could use [`UniqueKey`](https://github.com/zio/zio/blob/master/core/shared/src/main/scala/zio/internal/UniqueKey.scala) instead

Reply:
Currently I can't because I rely on Long inside `ReleaseKey` for ordering. But I would like to embed the ordering into the data structure holding the finalizers.

Reply:
Doe we maybe want to make it an abstract path-dependant type on ReleaseMap to make sure that 1. users can't do anything with it except use it as a key into the map 2. make sure they can't mix the keys

Reply:
Yes, good idea! Will do that.